src/core/qgsfeaturefiltermodel.cpp

```diff
@@ -0,0 +1,306 @@
+
+    int newEntriesSize = entries.size();
+
+    int currentEntryInNewList = -1;
+
+    for ( int i = 0; i < newEntriesSize; ++i )
+    {
+      if ( entries.at( i ).id == mCurrentFeatureId )
+      {
+        currentEntryInNewList = i;
+        break;
+      }
+    }
+
+    int firstRow = 0;
+
+    // Move the current entry to the first position
+    if ( mCurrentEntryIndex != -1 )
+    {
+      beginMoveRows( QModelIndex(), mCurrentEntryIndex, mCurrentEntryIndex, QModelIndex(), 0 );
+      mEntries.move( mCurrentEntryIndex, 0 );
+      endMoveRows();
+      firstRow = 1;
+    }
+
+    // Remove all outdated entries
+    beginRemoveRows( QModelIndex(), firstRow, mEntries.size() - firstRow );
+    mEntries.remove( firstRow, mEntries.size() - firstRow );
+    endRemoveRows();
+
+    // If we don't have a default entry yet, add one, if not available, it will be a QString()
+    if ( mEntries.isEmpty() )
+    {
+      QString currentDisplayExpression = mEntries.value( currentEntryInNewList ).value;
+      mEntries.append( Entry( mCurrentFeatureId, currentDisplayExpression ) );
+
+      if ( currentDisplayExpression.isNull() )
+        reloadCurrentFeature();
+    }
+
+    if ( currentEntryInNewList == -1 )
+    {
+      beginInsertRows( QModelIndex(), 1, entries.size() + 1 );
+      mEntries += entries;
+      endInsertRows();
+    }
+    else
+    {
+      if ( currentEntryInNewList != 0 )
+      {
+        beginInsertRows( QModelIndex(), 0, currentEntryInNewList - 1 );
+        mEntries = entries.mid( 0, currentEntryInNewList );
+        endInsertRows();
+      }
+      else
+      {
+        mEntries.replace( 0, entries.at( 0 ) );
+      }
+
+      emit dataChanged( index( currentEntryInNewList, 0, QModelIndex() ), index( currentEntryInNewList, 0, QModelIndex() ) );
+
+      beginInsertRows( QModelIndex(), currentEntryInNewList + 1, newEntriesSize - currentEntryInNewList - 1 );
+      mEntries += entries.mid( currentEntryInNewList + 1 );
+      endInsertRows();
+    }
+
+    emit filterJobCompleted();
+  }
+}
+
+void QgsFeatureFilterModel::gathererThreadFinished()
+{
+  delete mGatherer;
+  mGatherer = nullptr;
+  emit loadingChanged();
+}
+
+void QgsFeatureFilterModel::scheduledReload()
+{
+  bool wasLoading = false;
+
+  if ( mGatherer )
+  {
+    // Send the gatherer thread to the graveyard:
+    //   forget about it, tell it to stop and delete when finished
+    disconnect( mGatherer, &QgsFieldExpressionValuesGatherer::collectedValues, this, &QgsFeatureFilterModel::updateCompleter );
+    disconnect( mGatherer, &QgsFieldExpressionValuesGatherer::finished, this, &QgsFeatureFilterModel::gathererThreadFinished );
+    connect( mGatherer, &QgsFieldExpressionValuesGatherer::finished, mGatherer, &QgsFieldExpressionValuesGatherer::deleteLater );
+    mGatherer->stop();
+    wasLoading = true;
+  }
+
+  QgsFeatureRequest request;
+  if ( mShouldReloadCurrentFeature )
```

Comment:
You could get a boost by not fetching geometry if the display expression doesn't need it, and also setting a subset of attributes here based on the display expression